File: lint-rules-android-lint/src/main/java/com/vanniktech/lintrules/android/ShouldUseStaticImportDetector.java
package com.vanniktech.lintrules.android;

import com.android.annotations.NonNull;
import com.android.annotations.Nullable;
import com.android.tools.lint.detector.api.Detector;
import com.android.tools.lint.detector.api.Implementation;
import com.android.tools.lint.detector.api.Issue;
import com.android.tools.lint.detector.api.JavaContext;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiMethod;
import java.util.ArrayList;
import java.util.EnumSet;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.jetbrains.uast.UCallExpression;
import org.jetbrains.uast.UElement;
import org.jetbrains.uast.UFile;
import org.jetbrains.uast.UImportStatement;
import org.jetbrains.uast.UReferenceExpression;

import static com.android.tools.lint.detector.api.Category.CORRECTNESS;
import static com.android.tools.lint.detector.api.Scope.JAVA_FILE;
import static com.android.tools.lint.detector.api.Scope.TEST_SOURCES;
import static com.android.tools.lint.detector.api.Severity.WARNING;

@SuppressWarnings("checkstyle:multiplestringliterals") public class ShouldUseStaticImportDetector extends Detector implements Detector.UastScanner {
  private static final Map<String, String> METHODS_TO_STATICALLY_IMPORT = new HashMap<>();
  private static final Map<String, String> REFERENCES_TO_STATICALLY_IMPORT = new HashMap<>();

  static final Issue ISSUE_SHOULD_USE_STATIC_IMPORT =
      Issue.create("ShouldUseStaticImport", "Should be using a static import.",
          "Should be using a static import.", CORRECTNESS, 3,
          WARNING, new Implementation(ShouldUseStaticImportDetector.class, EnumSet.of(JAVA_FILE, TEST_SOURCES)));

  static {
    // Arrays.
    METHODS_TO_STATICALLY_IMPORT.put("asList", "java.util.Arrays");

    // Collections.
    METHODS_TO_STATICALLY_IMPORT.put("emptyEnumeration", "java.util.Collections");
    METHODS_TO_STATICALLY_IMPORT.put("emptyIterator", "java.util.Collections");
    METHODS_TO_STATICALLY_IMPORT.put("emptyList", "java.util.Collections");
    METHODS_TO_STATICALLY_IMPORT.put("emptyListIterator", "java.util.Collections");
    METHODS_TO_STATICALLY_IMPORT.put("emptyMap", "java.util.Collections");
    METHODS_TO_STATICALLY_IMPORT.put("emptySet", "java.util.Collections");
    METHODS_TO_STATICALLY_IMPORT.put("singleton", "java.util.Collections");
    METHODS_TO_STATICALLY_IMPORT.put("singletonList", "java.util.Collections");
    METHODS_TO_STATICALLY_IMPORT.put("singletonMap", "java.util.Collections");
    METHODS_TO_STATICALLY_IMPORT.put("singletonIterator", "java.util.Collections");
    METHODS_TO_STATICALLY_IMPORT.put("singletonSpliterator", "java.util.Collections");
    METHODS_TO_STATICALLY_IMPORT.put("unmodifiableCollection", "java.util.Collections");
    METHODS_TO_STATICALLY_IMPORT.put("unmodifiableSet", "java.util.Collections");
    METHODS_TO_STATICALLY_IMPORT.put("unmodifiableSortedSet", "java.util.Collections");
    METHODS_TO_STATICALLY_IMPORT.put("unmodifiableNavigableSet", "java.util.Collections");
    METHODS_TO_STATICALLY_IMPORT.put("unmodifiableList", "java.util.Collections");
    METHODS_TO_STATICALLY_IMPORT.put("unmodifiableMap", "java.util.Collections");
    METHODS_TO_STATICALLY_IMPORT.put("unmodifiableSortedMap", "java.util.Collections");
    METHODS_TO_STATICALLY_IMPORT.put("unmodifiableNavigableMap", "java.util.Collections");

    // Mockito.
    METHODS_TO_STATICALLY_IMPORT.put("mock", "org.mockito.Mockito.mock");
    METHODS_TO_STATICALLY_IMPORT.put("mockingDetails", "org.mockito.Mockito.mockingDetails");
    METHODS_TO_STATICALLY_IMPORT.put("spy", "org.mockito.Mockito.spy");
    METHODS_TO_STATICALLY_IMPORT.put("verify", "org.mockito.Mockito.verify");
    METHODS_TO_STATICALLY_IMPORT.put("reset", "org.mockito.Mockito.reset");
    METHODS_TO_STATICALLY_IMPORT.put("clearInvocations", "org.mockito.Mockito.clearInvocations");
    METHODS_TO_STATICALLY_IMPORT.put("verifyNoMoreInteractions", "org.mockito.Mockito.verifyNoMoreInteractions");
    METHODS_TO_STATICALLY_IMPORT.put("verifyZeroInteractions", "org.mockito.Mockito.verifyZeroInteractions");
    METHODS_TO_STATICALLY_IMPORT.put("when", "org.mockito.Mockito.when");
    METHODS_TO_STATICALLY_IMPORT.put("doThrow", "org.mockito.Mockito.doThrow");
    METHODS_TO_STATICALLY_IMPORT.put("doCallRealMethod", "org.mockito.Mockito.doCallRealMethod");
    METHODS_TO_STATICALLY_IMPORT.put("doAnswer", "org.mockito.Mockito.doAnswer");
    METHODS_TO_STATICALLY_IMPORT.put("doNothing", "org.mockito.Mockito.doNothing");
    METHODS_TO_STATICALLY_IMPORT.put("doReturn", "org.mockito.Mockito.doReturn");
    METHODS_TO_STATICALLY_IMPORT.put("inOrder", "org.mockito.Mockito.inOrder");
    METHODS_TO_STATICALLY_IMPORT.put("ignoreStubs", "org.mockito.Mockito.ignoreStubs");
    METHODS_TO_STATICALLY_IMPORT.put("never", "org.mockito.Mockito.never");
    METHODS_TO_STATICALLY_IMPORT.put("times", "org.mockito.Mockito.times");
    METHODS_TO_STATICALLY_IMPORT.put("atLeastOnce", "org.mockito.Mockito.atLeastOnce");
    METHODS_TO_STATICALLY_IMPORT.put("atLeast", "org.mockito.Mockito.atLeast");
    METHODS_TO_STATICALLY_IMPORT.put("atMost", "org.mockito.Mockito.atMost");
    METHODS_TO_STATICALLY_IMPORT.put("calls", "org.mockito.Mockito.calls");
    METHODS_TO_STATICALLY_IMPORT.put("only", "org.mockito.Mockito.only");
    METHODS_TO_STATICALLY_IMPORT.put("timeout", "org.mockito.Mockito.timeout");
    METHODS_TO_STATICALLY_IMPORT.put("after", "org.mockito.Mockito.after");
    METHODS_TO_STATICALLY_IMPORT.put("validateMockitoUsage", "org.mockito.Mockito.validateMockitoUsage");
    METHODS_TO_STATICALLY_IMPORT.put("withSettings", "org.mockito.Mockito.withSettings");
    METHODS_TO_STATICALLY_IMPORT.put("description", "org.mockito.Mockito.description");
    METHODS_TO_STATICALLY_IMPORT.put("debug", "org.mockito.Mockito.debug");
    METHODS_TO_STATICALLY_IMPORT.put("framework", "org.mockito.Mockito.framework");
    METHODS_TO_STATICALLY_IMPORT.put("mockitoSession", "org.mockito.Mockito.mockitoSession");

    // TimeUnit.
    REFERENCES_TO_STATICALLY_IMPORT.put("NANOSECONDS", "java.util.concurrent.TimeUnit");
    REFERENCES_TO_STATICALLY_IMPORT.put("MICROSECONDS", "java.util.concurrent.TimeUnit");
    REFERENCES_TO_STATICALLY_IMPORT.put("MILLISECONDS", "java.util.concurrent.TimeUnit");
    REFERENCES_TO_STATICALLY_IMPORT.put("SECONDS", "java.util.concurrent.TimeUnit");
    REFERENCES_TO_STATICALLY_IMPORT.put("MINUTES", "java.util.concurrent.TimeUnit");
    REFERENCES_TO_STATICALLY_IMPORT.put("HOURS", "java.util.concurrent.TimeUnit");
    REFERENCES_TO_STATICALLY_IMPORT.put("DAYS", "java.util.concurrent.TimeUnit");
    // Locales.
    REFERENCES_TO_STATICALLY_IMPORT.put("ENGLISH", "java.util.Locale");
    REFERENCES_TO_STATICALLY_IMPORT.put("FRENCH", "java.util.Locale");
    REFERENCES_TO_STATICALLY_IMPORT.put("GERMAN", "java.util.Locale");
    REFERENCES_TO_STATICALLY_IMPORT.put("ITALIAN", "java.util.Locale");
    REFERENCES_TO_STATICALLY_IMPORT.put("JAPANESE", "java.util.Locale");
    REFERENCES_TO_STATICALLY_IMPORT.put("KOREAN", "java.util.Locale");
    REFERENCES_TO_STATICALLY_IMPORT.put("CHINESE", "java.util.Locale");
    REFERENCES_TO_STATICALLY_IMPORT.put("SIMPLIFIED_CHINESE", "java.util.Locale");
    REFERENCES_TO_STATICALLY_IMPORT.put("TRADITIONAL_CHINESE", "java.util.Locale");
    REFERENCES_TO_STATICALLY_IMPORT.put("FRANCE", "java.util.Locale");
    REFERENCES_TO_STATICALLY_IMPORT.put("GERMANY", "java.util.Locale");
    REFERENCES_TO_STATICALLY_IMPORT.put("ITALY", "java.util.Locale");
    REFERENCES_TO_STATICALLY_IMPORT.put("JAPAN", "java.util.Locale");
    REFERENCES_TO_STATICALLY_IMPORT.put("KOREA", "java.util.Locale");
    REFERENCES_TO_STATICALLY_IMPORT.put("CHINA", "java.util.Locale");
    REFERENCES_TO_STATICALLY_IMPORT.put("PRC", "java.util.Locale");
    REFERENCES_TO_STATICALLY_IMPORT.put("TAIWAN", "java.util.Locale");
    REFERENCES_TO_STATICALLY_IMPORT.put("UK", "java.util.Locale");
    REFERENCES_TO_STATICALLY_IMPORT.put("US", "java.util.Locale");
    REFERENCES_TO_STATICALLY_IMPORT.put("CANADA", "java.util.Locale");
    REFERENCES_TO_STATICALLY_IMPORT.put("CANADA_FRENCH", "java.util.Locale");
    REFERENCES_TO_STATICALLY_IMPORT.put("ROOT", "java.util.Locale");
    // Android View.
    REFERENCES_TO_STATICALLY_IMPORT.put("VISIBLE", "android.view.View");
    REFERENCES_TO_STATICALLY_IMPORT.put("GONE", "android.view.View");
    REFERENCES_TO_STATICALLY_IMPORT.put("INVISIBLE", "android.view.View");
    // Android Service.
    REFERENCES_TO_STATICALLY_IMPORT.put("START_CONTINUATION_MASK", "android.app.Service");
    REFERENCES_TO_STATICALLY_IMPORT.put("START_STICKY_COMPATIBILITY", "android.app.Service");
    REFERENCES_TO_STATICALLY_IMPORT.put("START_STICKY", "android.app.Service");
    REFERENCES_TO_STATICALLY_IMPORT.put("START_NOT_STICKY", "android.app.Service");
    REFERENCES_TO_STATICALLY_IMPORT.put("START_REDELIVER_INTENT", "android.app.Service");
    REFERENCES_TO_STATICALLY_IMPORT.put("START_TASK_REMOVED_COMPLETE", "android.app.Service");
    REFERENCES_TO_STATICALLY_IMPORT.put("START_FLAG_REDELIVERY", "android.app.Service");
    REFERENCES_TO_STATICALLY_IMPORT.put("START_FLAG_RETRY", "android.app.Service");
    // Android Versions.
    REFERENCES_TO_STATICALLY_IMPORT.put("INCREMENTAL", "android.os.Build.VERSION");
    REFERENCES_TO_STATICALLY_IMPORT.put("RELEASE", "android.os.Build.VERSION");
    REFERENCES_TO_STATICALLY_IMPORT.put("BASE_OS", "android.os.Build.VERSION");
    REFERENCES_TO_STATICALLY_IMPORT.put("SECURITY_PATCH", "android.os.Build.VERSION");
    REFERENCES_TO_STATICALLY_IMPORT.put("SDK", "android.os.Build.VERSION");
    REFERENCES_TO_STATICALLY_IMPORT.put("SDK_INT", "android.os.Build.VERSION");
    REFERENCES_TO_STATICALLY_IMPORT.put("PREVIEW_SDK_INT", "android.os.Build.VERSION");
    REFERENCES_TO_STATICALLY_IMPORT.put("CODENAME", "android.os.Build.VERSION");
    REFERENCES_TO_STATICALLY_IMPORT.put("RESOURCES_SDK_INT", "android.os.Build.VERSION");
    // Android Version Codes.
    REFERENCES_TO_STATICALLY_IMPORT.put("CUR_DEVELOPMENT", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("BASE", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("BASE_1_1", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("CUPCAKE", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("DONUT", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("ECLAIR", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("ECLAIR_0_1", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("ECLAIR_MR1", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("FROYO", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("GINGERBREAD", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("GINGERBREAD_MR1", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("HONEYCOMB", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("HONEYCOMB_MR1", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("HONEYCOMB_MR2", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("ICE_CREAM_SANDWICH", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("ICE_CREAM_SANDWICH_MR1", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("JELLY_BEAN", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("JELLY_BEAN_MR1", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("JELLY_BEAN_MR2", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("KITKAT", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("KITKAT_WATCH", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("LOLLIPOP", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("LOLLIPOP_MR1", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("VERSION_CODES.M", "android.os.Build");
    REFERENCES_TO_STATICALLY_IMPORT.put("VERSION_CODES.N", "android.os.Build");
    REFERENCES_TO_STATICALLY_IMPORT.put("N_MR1", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("O", "android.os.Build.VERSION_CODES");
    // Future proof Version Codes.
    REFERENCES_TO_STATICALLY_IMPORT.put("P", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("Q", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("R", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("S", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("T", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("U", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("V", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("W", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("X", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("Y", "android.os.Build.VERSION_CODES");
    REFERENCES_TO_STATICALLY_IMPORT.put("Z", "android.os.Build.VERSION_CODES");
    // Retention Policy.
    REFERENCES_TO_STATICALLY_IMPORT.put("SOURCE", "java.lang.annotation.RetentionPolicy");
    REFERENCES_TO_STATICALLY_IMPORT.put("CLASS", "java.lang.annotation.RetentionPolicy");
    REFERENCES_TO_STATICALLY_IMPORT.put("RUNTIME", "java.lang.annotation.RetentionPolicy");
    // Mockito.
    REFERENCES_TO_STATICALLY_IMPORT.put("STRICT_STUBS", "org.mockito.quality.Strictness");
    REFERENCES_TO_STATICALLY_IMPORT.put("WARN", "org.mockito.quality.Strictness");
    REFERENCES_TO_STATICALLY_IMPORT.put("LENIENT", "org.mockito.quality.Strictness");
  }

  @NonNull private static UFile getUFile(@Nullable final UElement element) {
    if (element instanceof UFile) {
      return (UFile) element;
    }

    if (element != null) {
      return getUFile(element.getUastParent());
    }

    throw new IllegalArgumentException("Can't get file from element");
  }

  @Override public List<String> getApplicableMethodNames() {
    return new ArrayList<>(METHODS_TO_STATICALLY_IMPORT.keySet());
  }

  @Override public void visitMethod(final JavaContext context, final UCallExpression node, final PsiMethod method) {
    final String methodName = node.getMethodName();
    final String className = METHODS_TO_STATICALLY_IMPORT.get(methodName);

    if (methodName != null && className != null) {
      final PsiElement psi = node.getPsi();
      final boolean isStaticallyImported = psi != null && psi.getText().startsWith(methodName);
      final boolean matches = context.getEvaluator().isMemberInClass(method, className);

      if (!isStaticallyImported && matches) {
        context.report(ISSUE_SHOULD_USE_STATIC_IMPORT, node, context.getNameLocation(node), "Should statically import " + methodName);
      }
    }
  }

  @Override public List<String> getApplicableReferenceNames() {
    return new ArrayList<>(REFERENCES_TO_STATICALLY_IMPORT.keySet());
  }

  @Override public void visitReference(final JavaContext context, final UReferenceExpression expression, final PsiElement referenced) {
    final String name = expression.asRenderString();
    final UFile uFile = getUFile(expression);

    boolean isStaticallyImported = false;

    for (final UImportStatement uImportStatement : uFile.getImports()) {
      if (uImportStatement.asSourceString().contains(name)) {
        isStaticallyImported = true;
        break;
      }
    }

    if (!isStaticallyImported) {
      context.report(ISSUE_SHOULD_USE_STATIC_IMPORT, expression, context.getLocation(expression), "Should statically import " + name);
    }
  }
}

File: lint-rules-android-lint/src/main/java/com/vanniktech/lintrules/android/ShouldUseStaticImportDetector.kt
package com.vanniktech.lintrules.android

import com.android.tools.lint.detector.api.Category.Companion.CORRECTNESS
import com.android.tools.lint.detector.api.Context
import com.android.tools.lint.detector.api.Detector
import com.android.tools.lint.detector.api.Implementation
import com.android.tools.lint.detector.api.Issue
import com.android.tools.lint.detector.api.JavaContext
import com.android.tools.lint.detector.api.Scope.JAVA_FILE
import com.android.tools.lint.detector.api.Severity.WARNING
import com.android.tools.lint.detector.api.UastLintUtils
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiMethod
import org.jetbrains.uast.UCallExpression
import org.jetbrains.uast.UQualifiedReferenceExpression
import org.jetbrains.uast.UReferenceExpression
import org.jetbrains.uast.getContainingUFile
import java.util.EnumSet

val ISSUE_SHOULD_USE_STATIC_IMPORT = Issue.create("ShouldUseStaticImport",
    "Flags declarations that should be statically imported.",
    "Certain declarations like TimeUnit.SECONDS should be statically imported to increase the readability.",
    CORRECTNESS, PRIORITY, WARNING,
    Implementation(ShouldUseStaticImportDetector::class.java, EnumSet.of(JAVA_FILE)))

@Suppress("Detekt.StringLiteralDuplication")
private val methodsToStaticallyImport = mapOf(
    // Arrays.
    "asList" to "java.util.Arrays",
    // Collections.
    "emptyEnumeration" to "java.util.Collections",
    "emptyIterator" to "java.util.Collections",
    "emptyList" to "java.util.Collections",
    "emptyListIterator" to "java.util.Collections",
    "emptyMap" to "java.util.Collections",
    "emptySet" to "java.util.Collections",
    "singleton" to "java.util.Collections",
    "singletonList" to "java.util.Collections",
    "singletonMap" to "java.util.Collections",
    "singletonIterator" to "java.util.Collections",
    "singletonSpliterator" to "java.util.Collections",
    "unmodifiableCollection" to "java.util.Collections",
    "unmodifiableSet" to "java.util.Collections",
    "unmodifiableSortedSet" to "java.util.Collections",
    "unmodifiableNavigableSet" to "java.util.Collections",
    "unmodifiableList" to "java.util.Collections",
    "unmodifiableMap" to "java.util.Collections",
    "unmodifiableSortedMap" to "java.util.Collections",
    "unmodifiableNavigableMap" to "java.util.Collections",
    // Mockito.
    "mock" to "org.mockito.Mockito.mock",
    "mockingDetails" to "org.mockito.Mockito.mockingDetails",
    "spy" to "org.mockito.Mockito.spy",
    "verify" to "org.mockito.Mockito.verify",
    "reset" to "org.mockito.Mockito.reset",
    "clearInvocations" to "org.mockito.Mockito.clearInvocations",
    "verifyNoMoreInteractions" to "org.mockito.Mockito.verifyNoMoreInteractions",
    "verifyZeroInteractions" to "org.mockito.Mockito.verifyZeroInteractions",
    "when" to "org.mockito.Mockito.when",
    "doThrow" to "org.mockito.Mockito.doThrow",
    "doCallRealMethod" to "org.mockito.Mockito.doCallRealMethod",
    "doAnswer" to "org.mockito.Mockito.doAnswer",
    "doNothing" to "org.mockito.Mockito.doNothing",
    "doReturn" to "org.mockito.Mockito.doReturn",
    "inOrder" to "org.mockito.Mockito.inOrder",
    "ignoreStubs" to "org.mockito.Mockito.ignoreStubs",
    "never" to "org.mockito.Mockito.never",
    "times" to "org.mockito.Mockito.times",
    "atLeastOnce" to "org.mockito.Mockito.atLeastOnce",
    "atLeast" to "org.mockito.Mockito.atLeast",
    "atMost" to "org.mockito.Mockito.atMost",
    "calls" to "org.mockito.Mockito.calls",
    "only" to "org.mockito.Mockito.only",
    "timeout" to "org.mockito.Mockito.timeout",
    "after" to "org.mockito.Mockito.after",
    "validateMockitoUsage" to "org.mockito.Mockito.validateMockitoUsage",
    "withSettings" to "org.mockito.Mockito.withSettings",
    "description" to "org.mockito.Mockito.description",
    "debug" to "org.mockito.Mockito.debug",
    "framework" to "org.mockito.Mockito.framework",
    "mockitoSession" to "org.mockito.Mockito.mockitoSession"
)

@Suppress("Detekt.StringLiteralDuplication")
private val referencesToStaticallyImport = mapOf(
    // TimeUnit.
    "NANOSECONDS" to "java.util.concurrent.TimeUnit",
    "MICROSECONDS" to "java.util.concurrent.TimeUnit",
    "MILLISECONDS" to "java.util.concurrent.TimeUnit",
    "SECONDS" to "java.util.concurrent.TimeUnit",
    "MINUTES" to "java.util.concurrent.TimeUnit",
    "HOURS" to "java.util.concurrent.TimeUnit",
    "DAYS" to "java.util.concurrent.TimeUnit",
    // Locales.
    "ENGLISH" to "java.util.Locale",
    "FRENCH" to "java.util.Locale",
    "GERMAN" to "java.util.Locale",
    "ITALIAN" to "java.util.Locale",
    "JAPANESE" to "java.util.Locale",
    "KOREAN" to "java.util.Locale",
    "CHINESE" to "java.util.Locale",
    "SIMPLIFIED_CHINESE" to "java.util.Locale",
    "TRADITIONAL_CHINESE" to "java.util.Locale",
    "FRANCE" to "java.util.Locale",
    "GERMANY" to "java.util.Locale",
    "ITALY" to "java.util.Locale",
    "JAPAN" to "java.util.Locale",
    "KOREA" to "java.util.Locale",
    "CHINA" to "java.util.Locale",
    "PRC" to "java.util.Locale",
    "TAIWAN" to "java.util.Locale",
    "UK" to "java.util.Locale",
    "US" to "java.util.Locale",
    "CANADA" to "java.util.Locale",
    "CANADA_FRENCH" to "java.util.Locale",
    "ROOT" to "java.util.Locale",
    // Android View.
    "VISIBLE" to "android.view.View",
    "GONE" to "android.view.View",
    "INVISIBLE" to "android.view.View",
    // Android Service.
    "START_CONTINUATION_MASK" to "android.app.Service",
    "START_STICKY_COMPATIBILITY" to "android.app.Service",
    "START_STICKY" to "android.app.Service",
    "START_NOT_STICKY" to "android.app.Service",
    "START_REDELIVER_INTENT" to "android.app.Service",
    "START_TASK_REMOVED_COMPLETE" to "android.app.Service",
    "START_FLAG_REDELIVERY" to "android.app.Service",
    "START_FLAG_RETRY" to "android.app.Service",
    // Android Versions.
    "INCREMENTAL" to "android.os.Build.VERSION",
    "RELEASE" to "android.os.Build.VERSION",
    "BASE_OS" to "android.os.Build.VERSION",
    "SECURITY_PATCH" to "android.os.Build.VERSION",
    "SDK" to "android.os.Build.VERSION",
    "SDK_INT" to "android.os.Build.VERSION",
    "PREVIEW_SDK_INT" to "android.os.Build.VERSION",
    "CODENAME" to "android.os.Build.VERSION",
    "RESOURCES_SDK_INT" to "android.os.Build.VERSION",
    // Android Build.
    "DISPLAY" to "android.os.Build",
    "PRODUCT" to "android.os.Build",
    "DEVICE" to "android.os.Build",
    "BOARD" to "android.os.Build",
    "CPU_ABI" to "android.os.Build",
    "CPU_ABI2" to "android.os.Build",
    "MANUFACTURER" to "android.os.Build",
    "BRAND" to "android.os.Build",
    "MODEL" to "android.os.Build",
    "BOOTLOADER" to "android.os.Build",
    "RADIO" to "android.os.Build",
    "HARDWARE" to "android.os.Build",
    "SERIAL" to "android.os.Build",
    // Android Version Codes.
    "CUR_DEVELOPMENT" to "android.os.Build.VERSION_CODES",
    "BASE" to "android.os.Build.VERSION_CODES",
    "BASE_1_1" to "android.os.Build.VERSION_CODES",
    "CUPCAKE" to "android.os.Build.VERSION_CODES",
    "DONUT" to "android.os.Build.VERSION_CODES",
    "ECLAIR" to "android.os.Build.VERSION_CODES",
    "ECLAIR_0_1" to "android.os.Build.VERSION_CODES",
    "ECLAIR_MR1" to "android.os.Build.VERSION_CODES",
    "FROYO" to "android.os.Build.VERSION_CODES",
    "GINGERBREAD" to "android.os.Build.VERSION_CODES",
    "GINGERBREAD_MR1" to "android.os.Build.VERSION_CODES",
    "HONEYCOMB" to "android.os.Build.VERSION_CODES",
    "HONEYCOMB_MR1" to "android.os.Build.VERSION_CODES",
    "HONEYCOMB_MR2" to "android.os.Build.VERSION_CODES",
    "ICE_CREAM_SANDWICH" to "android.os.Build.VERSION_CODES",
    "ICE_CREAM_SANDWICH_MR1" to "android.os.Build.VERSION_CODES",
    "JELLY_BEAN" to "android.os.Build.VERSION_CODES",
    "JELLY_BEAN_MR1" to "android.os.Build.VERSION_CODES",
    "JELLY_BEAN_MR2" to "android.os.Build.VERSION_CODES",
    "KITKAT" to "android.os.Build.VERSION_CODES",
    "KITKAT_WATCH" to "android.os.Build.VERSION_CODES",
    "LOLLIPOP" to "android.os.Build.VERSION_CODES",
    "LOLLIPOP_MR1" to "android.os.Build.VERSION_CODES",
    "VERSION_CODES.M" to "android.os.Build",
    "VERSION_CODES.N" to "android.os.Build",
    "N_MR1" to "android.os.Build.VERSION_CODES",
    "O" to "android.os.Build.VERSION_CODES",
    // Future proof Version Codes.
    "P" to "android.os.Build.VERSION_CODES",
    "Q" to "android.os.Build.VERSION_CODES",
    "R" to "android.os.Build.VERSION_CODES",
    "S" to "android.os.Build.VERSION_CODES",
    "T" to "android.os.Build.VERSION_CODES",
    "U" to "android.os.Build.VERSION_CODES",
    "V" to "android.os.Build.VERSION_CODES",
    "W" to "android.os.Build.VERSION_CODES",
    "X" to "android.os.Build.VERSION_CODES",
    "Y" to "android.os.Build.VERSION_CODES",
    "Z" to "android.os.Build.VERSION_CODES",
    // Retention Policy.
    "SOURCE" to "java.lang.annotation.RetentionPolicy",
    "CLASS" to "java.lang.annotation.RetentionPolicy",
    "RUNTIME" to "java.lang.annotation.RetentionPolicy",
    // Mockito.
    "STRICT_STUBS" to "org.mockito.quality.Strictness",
    "WARN" to "org.mockito.quality.Strictness",
    "LENIENT" to "org.mockito.quality.Strictness"
)

class ShouldUseStaticImportDetector : Detector(), Detector.UastScanner {
  override fun getApplicableMethodNames() = methodsToStaticallyImport.keys.toList()

  override fun visitMethodCall(context: JavaContext, node: UCallExpression, method: PsiMethod) {
    val methodName = node.methodName
    val className = methodsToStaticallyImport[methodName]

    if (methodName != null && className != null) {
      val isStaticallyImported = node.sourcePsi?.text?.startsWith(methodName) ?: false
      val matches = context.evaluator.isMemberInClass(method, className)

      if (!isStaticallyImported && matches) {
        context.report(ISSUE_SHOULD_USE_STATIC_IMPORT, node, context.getNameLocation(node), "Should statically import $methodName")
      }
    }
  }

  // For some reason visitReference may be called multiple times with the same parameters.
  private val processed = mutableSetOf<UReferenceExpression>()

  override fun getApplicableReferenceNames() = referencesToStaticallyImport.keys.toList()

  override fun visitReference(context: JavaContext, reference: UReferenceExpression, referenced: PsiElement) {
    if (processed.add(reference)) {
      val name = reference.asRenderString()
      val uFile = reference.getContainingUFile() ?: return

      for (uImportStatement in uFile.imports) {
        if (uImportStatement.asSourceString().contains(name)) {
          return // Static import and we don't need to look further.
        }
      }

      val isQualified = reference.uastParent is UQualifiedReferenceExpression
      val matches = UastLintUtils.getQualifiedName(referenced) == referencesToStaticallyImport[name] + "." + name

      if (matches && isQualified) {
        context.report(ISSUE_SHOULD_USE_STATIC_IMPORT, reference, context.getLocation(reference), "Should statically import $name")
      }
    }
  }

  override fun afterCheckFile(context: Context) {
    super.afterCheckFile(context)
    processed.clear()
  }
}
